docs/design-documents/features/storage/TDBStore/TDBStore_design.md

```diff
@@ -0,0 +1,497 @@
+- Release `_mutex`     
+- Return OK
+
+### Incremental set APIs
+
+**set_start function**
+
+Header:  
+`virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);`
+
+Pseudo code:
+- Take `_mutex`  
+- Check if final size fits in free space, if not call `garbage_collection`
+- Call `find_record` to find record in storage
+- If found and `int_flags` field in header includes write once flag, return "write once" error
+- Allocate an `inc_set_handle_t` structure into `handle`
+- Calculate hash on `key` and update in `handle`
+- Update `bd_base_offset` in handle to `_free_space_offset`
+- Update a `record_header_t` structure with all relevant values
+- Update `key_size`, `data_size` and `create_flags` in `handle`
+- Calculate crc on header
+- Update crc with key and update in `handle`
+- Program key in position after header
+- Advance `_free_space_offset` and update in `bd_curr_offset` field in handle
+- Set `_free_space_offset` and update in `bd_curr_offset` field in handle
+- Call `find_record` to calculate hash and find record in storage (with null key and current hash)
+- If not found, add entry in ram table
+- Update offset in ram table
+
+**set_add_data function**
+
+Header:  
+`virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);`
+
+Pseudo code:
+- Calculate crc on `value_data` and update in handle
+- Program `value_data` from `bd_curr_offset`
+- Advance `bd_curr_offset`
+
+**set_finalize function**
+
+Header:  
+`virtual int set_finalize(set_handle_t handle);`
+
+Pseudo code:
+- Advance `_free_space_offset` to padded offset 
+- Update a `record_header_t` structure with all relevant values
+- Program header at `bd_base_offset` from handle with pads
+- Call `sync` on buffered block device
+- Free `handle`
+- Release `_mutex`
+
+### Key iterator APIs
+
+**iterator_open function**
+
+Header:  
+`virtual int iterator_open(iterator_t *it, const char *prefix = NULL);`
+
+Pseudo code:
+- Take `_mutex`  
+- Allocate a `key_iterator_handle_t` structure into `it`
+- Set `ram_table_ind` field in iterator to 0
+- Duplicate `prefix` into same field in iterator
+- Release `_mutex`
+
+**iterator_next function**
+
+Header:  
+`virtual int iterator_next(iterator_t it, char *key, size_t key_size);`
+
+Pseudo code:
+- Take `_mutex`  
+- While `ram_table_ind` field in iterator smaller than `_num_keys`
+	- Read key pointed to by ram table in `ram_table_ind` into a local variable
+	- If name matches prefix
+		- Advance `ram_table_ind` field in iterator
+		- Copy name to `key` and return OK
+	- Advance `ram_table_ind` field in iterator 
+- Return "not found" error
+- Release `_mutex`
+
+**iterator_close function**
+
+Header:  
+`virtual int iterator_close(iterator_t it);`
+
+Pseudo code:
+- Release `prefix` field in iterator and structure allocated at `it`
+
+
+### Reserved space
+
+**reserved_space_set function**
+
+Header:  
+`virtual int reserved_space_set(void *data);`
```

Comment:
Out of curiousity what's the reserved space used for? Is it for system-level attributes?

Reply:
Ah, that makes sense. Is the reserved space at the end of the block? Does that mean we should be storing the ROT at the end of whichever block it resides in?

Interesting note: This same trick is used for FAT superblocks + MBR tables.

Reply:
Oh ok. I guess we should actually make the ROT configurable at the byte-level in case we want to hide it in the actual bootloader image.